lib/srv/discovery/access_graph.go

```diff
@@ -0,0 +1,304 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package discovery
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/credentials"
+
+	accessgraphv1alpha "github.com/gravitational/teleport/gen/proto/go/accessgraph/v1alpha"
+	"github.com/gravitational/teleport/lib/service/servicecfg"
+	aws_sync "github.com/gravitational/teleport/lib/srv/discovery/fetchers/aws-sync"
+)
+
+const (
+	// batchSize is the maximum number of resources to send in a single
+	// request to the access graph service.
+	batchSize = 500
+)
+
+func (s *Server) reconcileAccessGraph(ctx context.Context, currentTAGResources *aws_sync.Resources, stream accessgraphv1alpha.AccessGraphService_AWSEventsStreamClient) {
+	errG, ctx := errgroup.WithContext(ctx)
+	errG.SetLimit(3)
+	allFetchers := s.getAllAWSSyncFetchers()
+	errs := make([]error, 0, len(allFetchers))
+	results := make([]*aws_sync.Resources, 0, len(allFetchers))
+	resultsMu := sync.Mutex{}
+	collectResults := func(result *aws_sync.Resources, err error) {
+		resultsMu.Lock()
+		defer resultsMu.Unlock()
+		if err != nil {
+			errs = append(errs, err)
+		}
+		if result != nil {
+			results = append(results, result)
+		}
+	}
+	for _, fetcher := range allFetchers {
+		fetcher := fetcher
+		errG.Go(func() error {
+			result, err := fetcher.Poll(ctx)
+			collectResults(result, err)
+			return nil
+		})
+	}
+	// Wait for all fetchers to finish but don't return an error because
+	// we want to continue reconciling the access graph even if some
+	// fetchers fail.
+	_ = errG.Wait()
+	err := trace.NewAggregate(errs...)
+	if err != nil {
+		s.Log.WithError(err).Error("Error polling TAGs")
+	}
+	result := aws_sync.MergeResources(results...)
```

Comment:
Maybe this is a nit but I find that this section is fighting the way errgroups want to be used and then adding a mutex where channels would do fine.

> Package errgroup provides synchronization, error propagation, and Context cancelation for groups of goroutines working on subtasks of a common task.

You're going out of your way to avoid the error propagation and context cancelation, I think the idiomatic Go way to do this would be to just use channels.

```suggestion
	tokens := make(chan struct{}, 3)
	errs := make(chan error)
	results := make(chan *aws_sync.Resources)
	allFetchers := s.getAllAWSSyncFetchers()
	for _, fetcher := range allFetchers {
		<-tokens
		go func() {
			defer func() { tokens <- struct{}{} }()
			result, err := fetcher.Poll(ctx)
			if err != nil {
				errs <- trace.Wrap(err)
				return
			}
			results <- result
		}()
	}
	var allResults []*aws_sync.Resources
	for i := 0; i < len(allFetchers); i++ {
		select {
		case err := <-errs:
			// Continue reconciling the access graph even if some fetchers fail.
			s.Log.WithError(err).Warn("Error polling TAGs")
		case result := <-results:
			allResults = append(allResults, result)
		}
	}
	result := aws_sync.MergeResources(results...)
```

Reply:
> Maybe this is a nit but I find that this section is fighting the way errgroups want to be used and then adding a mutex where channels would do fine

I do agree about that but here it's a special case because `fetcher.Poll` can return both errors and result. You might have permissions to pull aws ec2 isntances but miss some RDS database permissions and to keep it syncing the cloud state to tag we should continue even if errors exist.

 I edited the code to match with your expectations